File: frontend/src/mocks/types/governance.ts
/** Governance domain types — Policies & KPIs */

export type PolicyStatus =
  | "Draft"
  | "Under Review"
  | "Pending Approval"
  | "Approved"
  | "Published"
  | "Expired"
  | "Archived";

export type ApprovalStatus =
  | "Not Submitted"
  | "Pending"
  | "Approved"
  | "Rejected"
  | "Changes Requested";

export type PolicyCategory =
  | "Information Security"
  | "Access Control"
  | "Risk Assessment"
  | "Data Protection"
  | "Business Continuity"
  | "Disaster Recovery"
  | "Third-Party Risk"
  | "Incident Management"
  | "Acceptable Use"
  | "Identity and Access Management"
  | "Cloud Security"
  | "Asset Management"
  | "Compliance";

export type GovernanceDepartment =
  | "Cybersecurity GRC"
  | "Cybersecurity Operations"
  | "Information Technology"
  | "Finance"
  | "Human Resources"
  | "Legal"
  | "Operations"
  | "Internal Audit"
  | "Risk Assessment"
  | "Compliance"
  | "Procurement"
  | "Business Continuity"
  | "Executive Management";

export type ReviewFrequency =
  | "Monthly"
  | "Quarterly"
  | "Semi-Annual"
  | "Annual"
  | "On Demand";

export interface PolicyVersion {
  version: string;
  changeSummary: string;
  changedBy: string;
  changeDate: string;
  approvalStatus: ApprovalStatus;
  isCurrent: boolean;
}

export interface PolicyActivity {
  id: string;
  at: string;
  actor: string;
  action: string;
  detail?: string;
}

export interface GovernancePolicy {
  id: string;
  name: string;
  description: string;
  category: PolicyCategory;
  department: GovernanceDepartment;
  owner: string;
  approver: string;
  version: string;
  effectiveDate: string;
  nextReviewDate: string;
  reviewFrequency: ReviewFrequency;
  approvalStatus: ApprovalStatus;
  policyStatus: PolicyStatus;
  frameworks: string[];
  controls: string[];
  documentName?: string;
  evidenceNames: string[];
  notes: string;
  lastUpdated: string;
  versions: PolicyVersion[];
  activity: PolicyActivity[];
}

export type KpiCategory =
  | "Governance"
  | "Compliance"
  | "Risk Assessment"
  | "Identity & Access"
  | "Security Operations"
  | "Business Continuity"
  | "Disaster Recovery"
  | "Third-Party Risk"
  | "Policy Management"
  | "Audit";

export type KpiFrequency =
  | "Daily"
  | "Weekly"
  | "Monthly"
  | "Quarterly"
  | "Semi-Annual"
  | "Annual"
  | "On Demand";

export type KpiUnit =
  | "Percentage"
  | "Count"
  | "Hours"
  | "Days"
  | "Minutes"
  | "Currency"
  | "Score"
  | "Ratio"
  | "Custom";

export type KpiDataSource =
  | "Manual Entry"
  | "Excel Import"
  | "API"
  | "SIEM"
  | "SOAR"
  | "EDR"
  | "IAM"
  | "PAM"
  | "Active Directory"
  | "Cloud Platform"
  | "Audit System"
  | "Custom Integration";

export type PerformanceDirection =
  | "Higher Is Better"
  | "Lower Is Better"
  | "Target Range";

export type KpiStatusLabel = "On Target" | "Warning" | "Critical" | "No Data";

export interface KpiMeasurement {
  id: string;
  periodStart: string;
  periodEnd: string;
  value: number | null;
  status: KpiStatusLabel;
  notes?: string;
  recordedBy: string;
  recordedAt: string;
}

export interface GovernanceKpi {
  id: string;
  name: string;
  description: string;
  category: KpiCategory;
  department: GovernanceDepartment;
  owner: string;
  frequency: KpiFrequency;
  unit: KpiUnit;
  formula: string;
  direction: PerformanceDirection;
  target: number;
  warningThreshold: number;
  criticalThreshold: number;
  /** Used when direction is Target Range */
  targetMin?: number;
  targetMax?: number;
  currentValue: number | null;
  status: KpiStatusLabel;
  dataSource: KpiDataSource;
  periodStart: string;
  periodEnd: string;
  lastUpdated: string;
  evidenceNames: string[];
  notes: string;
  measurements: KpiMeasurement[];
}

export const GOVERNANCE_DEPARTMENTS: GovernanceDepartment[] = [
  "Cybersecurity GRC",
  "Cybersecurity Operations",
  "Information Technology",
  "Finance",
  "Human Resources",
  "Legal",
  "Operations",
  "Internal Audit",
  "Risk Assessment",
  "Compliance",
  "Procurement",
  "Business Continuity",
  "Executive Management",
];

export const POLICY_CATEGORIES: PolicyCategory[] = [
  "Information Security",
  "Access Control",
  "Risk Assessment",
  "Data Protection",
  "Business Continuity",
  "Disaster Recovery",
  "Third-Party Risk",
  "Incident Management",
  "Acceptable Use",
  "Identity and Access Management",
  "Cloud Security",
  "Asset Management",
  "Compliance",
];

export const POLICY_STATUSES: PolicyStatus[] = [
  "Draft",
  "Under Review",
  "Pending Approval",
  "Approved",
  "Published",
  "Expired",
  "Archived",
];

export const APPROVAL_STATUSES: ApprovalStatus[] = [
  "Not Submitted",
  "Pending",
  "Approved",
  "Rejected",
  "Changes Requested",
];

export const REVIEW_FREQUENCIES: ReviewFrequency[] = [
  "Monthly",
  "Quarterly",
  "Semi-Annual",
  "Annual",
  "On Demand",
];

export const KPI_CATEGORIES: KpiCategory[] = [
  "Governance",
  "Compliance",
  "Risk Assessment",
  "Identity & Access",
  "Security Operations",
  "Business Continuity",
  "Disaster Recovery",
  "Third-Party Risk",
  "Policy Management",
  "Audit",
];

export const KPI_FREQUENCIES: KpiFrequency[] = [
  "Daily",
  "Weekly",
  "Monthly",
  "Quarterly",
  "Semi-Annual",
  "Annual",
  "On Demand",
];

export const KPI_UNITS: KpiUnit[] = [
  "Percentage",
  "Count",
  "Hours",
  "Days",
  "Minutes",
  "Currency",
  "Score",
  "Ratio",
  "Custom",
];

export const KPI_DATA_SOURCES: KpiDataSource[] = [
  "Manual Entry",
  "Excel Import",
  "API",
  "SIEM",
  "SOAR",
  "EDR",
  "IAM",
  "PAM",
  "Active Directory",
  "Cloud Platform",
  "Audit System",
  "Custom Integration",
];

export const PERFORMANCE_DIRECTIONS: PerformanceDirection[] = [
  "Higher Is Better",
  "Lower Is Better",
  "Target Range",
];

export const KPI_STATUS_LABELS: KpiStatusLabel[] = [
  "On Target",
  "Warning",
  "Critical",
  "No Data",
];

/** Validate threshold logic for a KPI definition. Returns error message or null. */
export function validateKpiThresholds(options: {
  direction: PerformanceDirection;
  target: number;
  warningThreshold: number;
  criticalThreshold: number;
  targetMin?: number;
  targetMax?: number;
}): string | null {
  const {
    direction,
    target,
    warningThreshold,
    criticalThreshold,
    targetMin,
    targetMax,
  } = options;
  if ([target, warningThreshold, criticalThreshold].some((n) => Number.isNaN(n))) {
    return "KPI thresholds must be valid numbers.";
  }
  if (direction === "Higher Is Better") {
    if (!(criticalThreshold <= warningThreshold && warningThreshold <= target)) {
      return "For Higher Is Better: Critical ≤ Warning ≤ Target.";
    }
  } else if (direction === "Lower Is Better") {
    if (!(target <= warningThreshold && warningThreshold <= criticalThreshold)) {
      return "For Lower Is Better: Target ≤ Warning ≤ Critical.";
    }
  } else {
    const min = targetMin ?? Math.min(target, warningThreshold);
    const max = targetMax ?? Math.max(target, warningThreshold);
    if (!(min <= max)) {
      return "For Target Range: minimum must be less than or equal to maximum.";
    }
  }
  return null;
}

File: backend/app/schemas/governance.py
"""Pydantic schemas for Governance policies and KPIs."""

from __future__ import annotations

from datetime import date, datetime

from pydantic import BaseModel, Field, field_validator


class PolicyCreate(BaseModel):
    policy_id: str = Field(min_length=1, max_length=64)
    name: str = Field(min_length=1, max_length=255)
    description: str | None = None
    category: str
    department: str
    owner: str
    approver: str | None = None
    current_version: str = "1.0"
    effective_date: date | None = None
    next_review_date: date | None = None
    review_frequency: str | None = "Annual"
    approval_status: str = "Not Submitted"
    policy_status: str = "Draft"
    notes: str | None = None
    frameworks: list[str] = Field(default_factory=list)
    controls: list[str] = Field(default_factory=list)

    @field_validator("policy_id", "name", "owner")
    @classmethod
    def strip_required(cls, v: str) -> str:
        cleaned = v.strip()
        if not cleaned:
            raise ValueError("Required field cannot be empty")
        return cleaned


class PolicyUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    category: str | None = None
    department: str | None = None
    owner: str | None = None
    approver: str | None = None
    current_version: str | None = None
    effective_date: date | None = None
    next_review_date: date | None = None
    review_frequency: str | None = None
    approval_status: str | None = None
    policy_status: str | None = None
    notes: str | None = None
    frameworks: list[str] | None = None
    controls: list[str] | None = None


class PolicyOut(BaseModel):
    id: str
    policy_id: str
    name: str
    description: str | None
    category: str
    department: str
    owner: str
    approver: str | None
    current_version: str
    effective_date: date | None
    next_review_date: date | None
    review_frequency: str | None
    approval_status: str
    policy_status: str
    notes: str | None
    document_filename: str | None
    frameworks: list[str] = Field(default_factory=list)
    controls: list[str] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class PolicyActionRequest(BaseModel):
    comment: str | None = Field(default=None, max_length=1000)


class KpiCreate(BaseModel):
    kpi_id: str = Field(min_length=1, max_length=64)
    name: str = Field(min_length=1, max_length=255)
    description: str | None = None
    category: str
    department: str
    owner: str
    frequency: str = "Monthly"
    unit: str = "Percentage"
    formula: str | None = None
    performance_direction: str = "Higher Is Better"
    target: float
    warning_threshold: float
    critical_threshold: float
    target_min: float | None = None
    target_max: float | None = None
    data_source: str | None = "Manual Entry"
    notes: str | None = None

    @field_validator("kpi_id", "name", "owner")
    @classmethod
    def strip_required(cls, v: str) -> str:
        cleaned = v.strip()
        if not cleaned:
            raise ValueError("Required field cannot be empty")
        return cleaned


class KpiUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    category: str | None = None
    department: str | None = None
    owner: str | None = None
    frequency: str | None = None
    unit: str | None = None
    formula: str | None = None
    performance_direction: str | None = None
    target: float | None = None
    warning_threshold: float | None = None
    critical_threshold: float | None = None
    target_min: float | None = None
    target_max: float | None = None
    data_source: str | None = None
    notes: str | None = None
    is_active: bool | None = None


class KpiOut(BaseModel):
    id: str
    kpi_id: str
    name: str
    description: str | None
    category: str
    department: str
    owner: str
    frequency: str
    unit: str
    formula: str | None
    performance_direction: str
    target: float
    warning_threshold: float
    critical_threshold: float
    target_min: float | None
    target_max: float | None
    data_source: str | None
    is_active: bool
    notes: str | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class KpiMeasurementCreate(BaseModel):
    period_start: date | None = None
    period_end: date | None = None
    value: float | None = None
    notes: str | None = None


class KpiMeasurementOut(BaseModel):
    id: str
    period_start: date | None
    period_end: date | None
    value: float | None
    calculated_status: str | None
    notes: str | None
    recorded_by: str | None
    recorded_at: datetime

    model_config = {"from_attributes": True}
